### PR TITLE
Fixes #10: Add Dockerfile for Yacy

### LINKED ISSUE #10: Create a docker file for the mcp and attaches services
The docker file shall start the mcp, rabbitmq, ftp and elasticsearch in one docker instance.

### PR DESCRIPTION
**Pull request by harshit98**
Fixes #10 

In this PR: 
- Created a Dockerfile for running MCP service.
- Created a `docker-compose.yml` to link services: apache ftp, elasticsearch and rabbitmq containers.
- Configurations for apache ftp and rabbitmq are still pending to be written.

Kindly review @nikhilrayaprolu @Marauderer97 @Orbiter

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+#install gradle required for build
+RUN add-apt-repository ppa:cwchien/gradle
+RUN apt-get update
+RUN apt-get install gradle
+
+# install apache ftp server 1.1.0
+RUN wget http://www-eu.apache.org/dist/mina/ftpserver/1.1.0/dist/apache-ftpserver-1.1.0.tar.gz
+RUN tar xfz apache-ftpserver-1.1.0.tar.gz
+
```
Comment (nikhilrayaprolu) [inline Dockerfile]: configure apache ftp server
**Things to do:**

1.add these lines to apache-ftpserver-1.1.0/res/conf/users.properties
```
ftpserver.user.yacygrid.userpassword=fcd6f2c8d772db634cd9fd960f4668a0
ftpserver.user.yacygrid.homedirectory=./res/home
ftpserver.user.yacygrid.enableflag=true
ftpserver.user.yacygrid.writepermission=true
ftpserver.user.yacygrid.maxloginnumber=2000
ftpserver.user.yacygrid.maxloginperip=2000
ftpserver.user.yacygrid.idletime=300
ftpserver.user.yacygrid.uploadrate=0
ftpserver.user.yacygrid.downloadrate=0
```

you can refer https://unix.stackexchange.com/questions/77277/how-to-append-multiple-lines-to-a-file for how to add multiple lines into a file.
but I prefer write these configurations into a configfile and then
do,

`cat ftpconfigfile >> apache-ftpserver-1.1.0/res/conf/users.properties`

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+RUN tar xf rabbitmq-server-generic-unix-3.6.6.tar.xz
+
+# run the RabbitMQ server
+RUN cd rabbitmq_server-3.6.6/sbin/rabbitmq-server
+
+# install erlang language for RabbitMQ
+RUN apt-get install erlang
+
+# install the management plugin to be able to use a web interface
+RUN rabbitmq_server-3.6.10/sbin/rabbitmq-plugins enable rabbitmq_management
```
Comment (nikhilrayaprolu) [inline Dockerfile]: `rabbitmq_server-3.6.6/sbin/rabbitmqctl add_user yacygrid password4account`
use the same username and password as given in the instructions

Comment (nikhilrayaprolu) [reply Dockerfile]: add these lines
`[{rabbit, [{loopback_users,[]}]}].` to rabbitmq_server-3.6.6/etc/rabbitmq/rabbitmq.config using `echo` command .

Comment (nikhilrayaprolu) [reply Dockerfile]: now configure MCP with the following lines
make a `MCPconfigfile`  file with same content as `/conf/config.properties` replace with these lines
```

grid.ftp.address = yacygrid:password4account@localhost:2121
grid.broker.address = yacygrid:password4account@localhost:5672
grid.elasticsearch.address = elasticsearch:9300
grid.elasticsearch.clusterName =yacygrid

```

**Review by nikhilrayaprolu — CHANGES_REQUESTED** [review]

File: docker-compose.yml [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+    build: .
+    ports:
+      - "80:80"
+    links:
+      - elasticsearch
+    environment:
+      ELASTIC_URI: elasticsearch://localhost:5050/yacy_grid_mcp
+
+  elasticsearch:
+    image: elasticsearch
```
Comment (nikhilrayaprolu) [inline docker-compose.yml]: replace this with compose format given at https://www.elastic.co/guide/en/elasticsearch/reference/current/docker.html

Comment (nikhilrayaprolu) [reply docker-compose.yml]: also update the clustername to "yacygrid" as given in instructions

**Review by nikhilrayaprolu — CHANGES_REQUESTED** [review]

**Comment by harshit98** [comment]
@nikhilrayaprolu Kindly review.

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+# use the same username and password as given in instructions
+RUN rabbitmq_server-3.6.6/sbin/rabbitmqctl add_user yacygrid password4account
+RUN echo [{rabbit, [{loopback_users, []}]}].
```
Comment (nikhilrayaprolu) [inline Dockerfile]: add this line into `rabbitmq_server-3.6.6/etc/rabbitmq/rabbitmq.config`

File: config-mcp.properties [hunk 5]
```diff
@@ -0,0 +1,28 @@
+# YaCy Grid Peer configuration
```
Comment (nikhilrayaprolu) [inline config-mcp.properties]: copy this configuration file to data/mcp-8100/conf/config.properties
using cp command

**Review by nikhilrayaprolu — CHANGES_REQUESTED** [review]

File: docker-compose.yml [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+        soft: -1
+        hard: -1
+    mem_limit: 1g
+    volumes:
+      - esdata1:/usr/share/elasticsearch/data
+    ports:
+      - 9200:9200
+    networks:
+      - esnet
+  elasticsearch2:
```
Comment (nikhilrayaprolu) [inline docker-compose.yml]: why do we need two elasticsearch containers.

**Review by nikhilrayaprolu — CHANGES_REQUESTED** [review]

File: docker-compose.yml [hunk 7]
```diff
@@ -0,0 +1,45 @@
+version: '2'
+services:
```
Comment (nikhilrayaprolu) [inline docker-compose.yml]:
```
version: "2"
  services:
  yacy_MCP:
   name: "yacy"
   build: .
   ports:
   - "80:80"
   links:
   - elasticsearch
 

```
missing the main service

**Review by nikhilrayaprolu — CHANGES_REQUESTED** [review]

File: docker-compose.yml (hunk 6 above)
Comment (harshit98) [reply docker-compose.yml]: Okay I will remove second container.

File: config-mcp.properties (hunk 5 above)
Comment (harshit98) [reply config-mcp.properties]: Done @nikhilrayaprolu

**Review by nikhilrayaprolu — CHANGES_REQUESTED** [review]
we have many files that are specifically for Docker, don't pull all these files in the root directory, create a folder named Docker and put all these files there. for instance check https://github.com/yacy/yacy_search_server

**Review by Orbiter — APPROVED** [review]

**Comment by Orbiter** [comment]
yes looks good but please consider suggestion from @nikhilrayaprolu

**Comment by harshit98** [comment]
Done. @Orbiter @nikhilrayaprolu

**Comment by nikhilrayaprolu** [comment]
@harshit98 there are alot more changes needed to be done.

**Comment by nikhilrayaprolu** [comment]
I am testing my Docker on my local system will update you, once it is done.

**Comment by nikhilrayaprolu** [comment]
@harshit98 you can close this pull request or update it with the changes I made at https://github.com/yacy/yacy_grid_mcp/pull/13

**Pull request closed without merging**